### PR TITLE
Feature/improve registerplanning more efficient data structures

### PR DESCRIPTION
**Pull request by goedderz**

### CODE REVIEW

**Comment by goedderz** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/9790/

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/9844/

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/9867/

File: CHANGELOG [hunk 1]
```diff
@@ -1,7 +1,10 @@
 devel
 -----
-* Moved calculation of registers to keep into the register planning. It can be
-  calculated more efficiently there.
+
+* Made several improvements to AQL's internal register planning. This will
+  improve query performance in several cases; mainly of large queries with many
+  variables, and of spliced subqueries. This also fixes the github issues #10853
+  and #11358.
```
Comment (Simran-B) [inline CHANGELOG:7]:
```suggestion
  and #11358 which reported performance degradation caused by subquery splicing.
```

Comment (goedderz) [reply CHANGELOG:7]: 👍

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/9874/

File: arangod/Aql/VarUsageFinder.cpp [hunk 2]
```diff
@@ -16,6 +19,18 @@ using namespace arangodb::aql;
...
 
+namespace {
+
+/// @brief Merge everything from source into target. (emilib::HashSet does not
+///        have a merge() method, nor extract()).
+auto mergeHashSets(VarSet& target, VarSet&& source) {
+  for (auto varPtr : source) {
+    target.emplace(varPtr);
+  }
+  source.clear();
```
Comment (markuspf) [inline arangod/Aql/VarUsageFinder.cpp:30]: Is this "standard" for `merge` operations?

Not only woudl I find it a bit weird that `mergeHashSets` merges the second argument into the first, but also that the second would then afterwards be empty. Maybe this is really just a naming issue

File: utils/bison-c.sh [hunk 3]
```diff
@@ -9,7 +9,7 @@ if test "x$BISON" = x -o "x$OUTPUT" = x -o "x$INPUT" = x;  then
   exit 1
 fi
 
-BISON_MAJOR_VER=`${BISON} --version |grep bison|sed -e "s;.* ;;" -e "s;\..*;;"`
+BISON_MAJOR_VER=`"${BISON}" --version |grep bison|sed -e "s;.* ;;" -e "s;\..*;;"`
```
Comment (markuspf) [inline utils/bison-c.sh:12]: Huh? How are these changes related?

Comment (goedderz) [reply utils/bison-c.sh:12]: They're only related in the sense that I couldn't compile, because my path to bison contains spaces. I can move them to a separate PR.

File: arangod/Aql/VarUsageFinder.cpp (hunk 2 above)
Comment (goedderz) [reply arangod/Aql/VarUsageFinder.cpp:30]: This is meant to replace
```
target.merge(std::move(source));
```
, that's where this is coming from. However, the `emilib::HashSet` does not have `merge` implemented. I'll extend this comment and add one at the caller's site.

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/9899/

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/9901/
